w3af/plugins/audit/hpp.py

```diff
@@ -0,0 +1,259 @@
+        self.parser = etree.HTMLParser(remove_blank_text=True, remove_comments=True)
+        self.precedence = None
+        self.rlink = ''
+
+
+    def audit(self, freq, orig_response):
+        """
+        Tests an URL for HPP vulnerabilities.
+
+        :param freq: A FuzzableRequest
+
+        """
+
+        rparams = {}
+
+        if not orig_response.is_text_or_html():
+            return
+
+        if not orig_response.get_code() == 200:
+            return
+        method = freq.get_method()
+        self._orig_uri = freq.get_uri()
+        if method == 'GET':
+            qs = freq.get_querystring()
+
+
+        #finding reflected params
+        for param in qs:
+            value = qs[param][0]
+            mutants = create_mutants(freq, [REFLECT_TOKEN],
+                                     fuzzable_param_list= [param],
+                                     mutant_tuple=(QSMutant,))
+            mu = mutants[0]
+            #first we send REFLECT_TOKEN
+            resp = self._uri_opener.send_mutant(mu)
+            try:
+                dp = parser_cache.dpc.get_document_parser_for(resp)
+            except BaseFrameworkException:
+                return
+            #do we have a reflection?
+
+            body = self.strip_trash(resp.body)
+            if not resp.get_code() == 200:
+                #no need to seek reflection if it is not OK
+                continue
+            rp = self.find_reflection( dp, REFLECT_TOKEN)
+
+            if not rp:
+                # no reflection at all skip the parameter
+                continue
+            rparams[param] = rp
+            # print "querystring parameter %s is reflected inside " \
+            #       "a link HTML code" % param
+            if self.precedence:
+               continue
+
+            #this hack is to make w3af send request with unencoded qs
+            u =  str(mu.get_uri())
+            ptest_url = URL(u.replace(REFLECT_TOKEN,
+                                      value +'&'+param+'='+REFLECT_TOKEN))
+            ptest_resp = self._uri_opener.GET(ptest_url)
+            ptest_resp_code = ptest_resp.get_code()
+            
+            if not ptest_resp_code == 200:
+                om.out.debug(u"Precedence check request to the uri %s "
+                             u"failed! "
+                             u"Got response code %s in response id %s\n"
+                             u"Unable to continue HPP detection. "
+                             u"Resource might be protected by a WAF,"
+                             % (ptest_url, ptest_resp_code, ptest_resp.id))
+                #no need to seek reflection if it is not OK
+                continue
+
+            ptest_body = self.strip_trash(ptest_resp.body)
+
+            try:
+                ptest_dp = parser_cache.dpc.get_document_parser_for(ptest_resp)
+            except BaseFrameworkException:
+                return
+
+            if fuzzy_equal(self.strip_trash(orig_response.body),
```

Comment:
`self.strip_trash(orig_response.body)` which is a CPU intensive task is calculated once for each QS parameter, move this outside the for loop